New option

ivan-jukic opened the pull request:
Hey @jmackie , hope all is good :)

I've made a small change on the `elm-smuggle`. I've added a new argument option `--elm-bin` which is for the purpose of passing a relative path to the elm binary that we want to use, instead of relying on the system wide elm installation. So if you have for e.g. 0.18 installed on the system, but have included 0.19 in your `packages.json`, you can now tell `elm-smuggle` here's the elm binary.

Tested it on Win and Linux, compiles and seems to be working. Please check my Haskell code, if it's at all close to your standards, and let me know what I need to fix.

Hopefully, if it's all good, we could update `elm-smuggle` on the NPM as well.

Thanks!

ivan-jukic committed:
Added --local-bin flag

ivan-jukic committed:
WIP fixing --local-bin flag

ivan-jukic committed:
Added --elm-bin flag
---
Added new flag to indicate relative location of the elm binary.
This way, we can use elm binary installed via npm, and do not
need to depend on the system wide installation of elm.

E.g. usage linux
elm-smuggle --elm-bin=node_modules/.bin/elm

E.g. usage windows
elm-smuggle --elm-bin=node_modules/elm/unpacked_bin/elm.exe

Also, because of differences betwen win and linux, this way we
can handle both platforms.

jmackie commented:
Hey man, cheers for this. The code looks good, but I do wonder if it's necessary? If you're using  `elm-smuggle` from npm, then could you not just `npx elm-smuggle` (or equivalent), that way the compiler binary will resolve to that in your `node_modules`?

ivan-jukic commented:
Thanks for the reply man! If you call `npx elm-smuggle`, and then `elm-smuggle` calls `elm make` command, wouldn't that still resolve to the system installation of elm?
Or you're saying that it might be easier if the flag would indicate that `elm-smuggle` should call `npx elm make`, instead of accepting a relative path to the elm binary?

jmackie commented:
`elm-smuggle` is just gonna call the `elm` binary highest up on your PATH.
So if you invoke it through npx, or by npm running a script in your
package.json, it's gonna pick up the one in your node_modules/.bin folder

On Fri, 29 Mar 2019, 17:37 Ivan Jukic, <notifications@github.com> wrote:

> Thanks for the reply man! If you call npx elm-smuggle, and then
> elm-smuggle calls elm make command, wouldn't that still resolve to the
> system installation of elm?
> Or you're saying that it might be easier if the flag would indicate that
> elm-smuggle should call npx elm make, instead of accepting a relative
> path to the elm binary?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmackie/elm-smuggle/pull/1#issuecomment-478086423>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALnwGxpEljnsMLhiwDHOn58VdaCKfsh7ks5vbk9EgaJpZM4cFCvi>
> .
>

ivan-jukic commented:
Sorry, yes you're right, it works well on Linux. Unfortunately, in this case my perspective is from Windows platform, and at the moment it doesn't seem like the `npx` approach works there. Could you please double check that, and confirm?

To test this I've removed `elm` and `elm-smuggle` from my pc, added both as dev dependencies to a dummy project, and ran `npx elm-smuggle` in the project dir, but on Windows it just returns `Error: elm not found`. It worked well in `WSL` tho, same procedure.

![elm_smuggle_win_vs_linux](https://user-images.githubusercontent.com/3020399/55255797-5dbde500-5253-11e9-8416-9044aa4c8de8.jpg)

P.S. it doesn't seem like that last change with version limits in `.elm-smuggle` got into `npm` :)

jmackie commented:
That makes all the sense. Classic Windows.

I'll try and figure out what's going on but I'm slightly limited by not
having access to a Windoze machine. Maybe `Path.IO.findExecutable` isn't
super portable? Or maybe npx isn't doing what I think it is (i.e. running
the process with a modified env)?


On Fri, 29 Mar 2019, 18:50 Ivan Jukic, <notifications@github.com> wrote:

> Sorry, yes you're right, it works well on Linux. Unfortunately, in this
> case my perspective is from Windows platform, and at the moment it doesn't
> seem like the npx approach works there. Could you please double check
> that, and confirm?
>
> To test this I've removed elm and elm-smuggle from my pc, added both as
> dev dependencies to a dummy project, and ran npx elm-smuggle in the
> project dir, but on Windows it just returns Error: elm not found. It
> worked well in WSL tho, same procedure.
>
> [image: elm_smuggle_win_vs_linux]
> <https://user-images.githubusercontent.com/3020399/55255797-5dbde500-5253-11e9-8416-9044aa4c8de8.jpg>
>
> P.S. it doesn't seem like that last change with version limits in
> .elm-smuggle got into npm :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmackie/elm-smuggle/pull/1#issuecomment-478110426>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALnwG-W0MSdQz4LuJp2bfdE3SSGC8E39ks5vbmCFgaJpZM4cFCvi>
> .
>

ivan-jukic commented:
Would this help/work https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/?

Btw, could you also check cases when elm-smuggle runs on Windows, but `elm cache` dir doesn't exist. Seems like it just hangs in the `Command.runWithYes` function. I tried fixing it in this commit https://github.com/ivan-jukic/elm-smuggle/commit/3406e7d201fe8cc8ed6998d1e912898cbbd581ec, tho I'm not a fan of the solution. Didn't had enough time to figure out what the exact issue was, but it seems related to the `Process.runInteractiveProcess` and Windows. Works fine on Linux obviously.
